pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,850 @@
+			// soon be deleted. The build should be transitioned to the Error phase.
+			if pod.DeletionTimestamp != nil {
+				return podDeletedUpdate(), nil
+			}
+
+			update.setPhase(buildapi.BuildPhaseFailed)
+			update.setReason(buildapi.StatusReasonGenericBuildFailed)
+			update.setMessage(buildapi.StatusMessageGenericBuildFailed)
+		}
+	}
+	return update, nil
+}
+
+// handleCompletedBuild will only be called on builds that are already in a terminal phase however, their completion timestamp
+// has not been set.
+func (bc *BuildController) handleCompletedBuild(build *buildapi.Build, pod *kapi.Pod, podErr error) (*buildUpdate, error) {
+	// Make sure that the completion timestamp has not already been set
+	if build.Status.CompletionTimestamp != nil {
+		return nil, nil
+	}
+
+	update := &buildUpdate{}
+	var podStartTime *metav1.Time
+	if podErr != nil && pod != nil {
+		podStartTime = pod.Status.StartTime
+	}
+	setBuildCompletionTimestampAndDuration(build, podStartTime, update)
+
+	return update, nil
+}
+
+// updateBuild is the single place where any update to a build is done in the build controller.
+// It will check that the update is valid, peform any necessary processing such as calling HandleBuildCompletion,
+// and apply the buildUpdate object as a patch.
+func (bc *BuildController) updateBuild(build *buildapi.Build, update *buildUpdate, pod *kapi.Pod) {
+
+	stateTransition := false
+	// Check whether we are transitioning to a different build phase
+	if update.phase != nil && (*update.phase) != build.Status.Phase {
+		stateTransition = true
+
+		// Make sure that the transition is valid
+		if !isValidTransition(build.Status.Phase, *update.phase) {
+			err := fmt.Errorf("invalid phase transition %s -> %s", buildDesc(build), *update.phase)
+			utilruntime.HandleError(err)
+			return
+		}
+
+		// Log that we are updating build status
+		reasonText := ""
+		if update.reason != nil && *update.reason != "" {
+			reasonText = fmt.Sprintf(" ( %s )", *update.reason)
+		}
+
+		// Update build completion timestamp if transitioning to a terminal phase
+		if isTerminal(*update.phase) {
+			var podStartTime *metav1.Time
+			if pod != nil {
+				podStartTime = pod.Status.StartTime
+			}
+			setBuildCompletionTimestampAndDuration(build, podStartTime, update)
+		}
+
+		glog.V(4).Infof("Updating build %s -> %s%s", buildDesc(build), *update.phase, reasonText)
+	}
+
+	// Treat a failed->failed update as a state transition when the completionTime is getting
+	// updated. This will cause an event to be emitted and completion processing to trigger.
+	if !stateTransition && build.Status.Phase == buildapi.BuildPhaseFailed && update.completionTime != nil {
+		glog.V(4).Infof("Processing build %s completion", buildDesc(build))
+		stateTransition = true
+	}
+
+	// Ensure that a pod name annotation has been set on the build if a pod is available
+	if update.podNameAnnotation == nil && !common.HasBuildPodNameAnnotation(build) && pod != nil {
```

Comment:
So we will set the podName annotation when we move the build from new to pending. However, if the update to pending failed, then we may not have pod name annotation on the build. This ensures that if there's a pod, we will have a pod name annotation. I can see that it can be moved to handleActiveBuild. Would that work?